### PR TITLE
Import / Export Schematics from JSON Files

### PR DESCRIPTION
**Pull request by pjdotson**
# Feature Pull Request Template

## Key Information

- **Linear Issue**: [SY-865](https://linear.app/synnaxlabs/issue/SY-865/[console]-import-and-export-schematic-from-json)

## Description

Added the ability to download a schematic as a JSON file and upload them back into the console.

## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added relevant tests to cover the changes.
- [x] I have updated user facing documentation accordingly.

## Migrations

- [x] Console - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.
- [x] Server - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.

## Manual QA Additions

- [x] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
with necessary manual QA steps to test my change.

## Breaking Changes

None

### CODE REVIEW

File: console/src/layouts/Mosaic.tsx [hunk 1]
```diff
@@ -56,20 +58,26 @@ export const Mosaic = memo((): ReactElement => {
...
   const store = useStore();
   const activeTab = useSelectActiveMosaicTabKey();
-
   const client = Synnax.use();
   const placer = usePlacer();
   const dispatch = useDispatch();
 
   const handleDrop = useCallback(
     (key: number, tabKey: string, loc: location.Location): void => {
+      console.log("handleDrop");
```
Comment (emilbon99) [inline console/src/layouts/Mosaic.tsx:67]: A bunch of console logs that need to be removed

File: console/src/main.tsx [hunk 2]
```diff
@@ -79,6 +79,11 @@ const MainUnderContext = (): ReactElement => {
   const theme = Layout.useThemeProvider();
   const cluster = Cluster.useSelect();
   const activeRange = Range.useSelect();
+  useEffect(() => {
```
Comment (emilbon99) [inline console/src/main.tsx:82]: We should extract this into a named hook or add a comment explaining why we put it there.

File: console/src/schematic/migrations/v2.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// Source License, use of this software will be governed by the Apache License,
+// Version 2.0, included in the file licenses/APL.txt.
+
+import { migrate } from "@synnaxlabs/x";
+import { z } from "zod";
+
+import * as v1 from "@/schematic/migrations/v1";
+
+export const stateZ = v1.stateZ.omit({ version: true }).extend({
+  version: z.literal("2.0.0"),
```
Comment (emilbon99) [inline console/src/schematic/migrations/v2.ts:16]: What about adding a type to differentiate it from a plot?

File: console/src/schematic/selectors.ts [hunk 4]
```diff
@@ -85,11 +85,12 @@ export const selectNodeProps = (
   return schematic.props[key];
 };
 
-export const useSelectNodeProps = (layoutKey: string, key: string): NodeProps =>
-  useMemoSelect(
+export const useSelectNodeProps = (layoutKey: string, key: string): NodeProps => {
+  return useMemoSelect(
```
Comment (emilbon99) [inline console/src/schematic/selectors.ts:89]: Reverse this - doesn't need brackets

File: console/src/schematic/services/file.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import { deep, errors, type UnknownRecord } from "@synnaxlabs/x";
+
+import { Layout } from "@/layout";
+import { moveMosaicTab } from "@/layout/slice";
+import { parser } from "@/schematic/migrations";
+import { create } from "@/schematic/Schematic";
+import { select } from "@/schematic/selectors";
+import { remove } from "@/schematic/slice";
+
+export const FileHandler: Layout.FileHandler = async ({
```
Comment (emilbon99) [inline console/src/schematic/services/file.ts:20]: Fix case -> `fileHandler`

File: console/src/schematic/services/file.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  if (client != null) {
+    try {
+      await client.workspaces.schematic.retrieve(key);
+      await client.workspaces.schematic.setData(
+        key,
+        newState as unknown as UnknownRecord,
+      );
+    } catch (e) {
+      if (!NotFoundError.matches(e)) throw e;
+      if (workspaceKey == null) throw Error("Workspace key is required.");
```
Comment (emilbon99) [inline console/src/schematic/services/file.ts:63]: Why throw here? Shouldn't we be able to load a schematic even if there is no active workspace?

File: console/src/schematic/services/file.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      await client.workspaces.schematic.create(workspaceKey, {
+        name,
+        data: newState as unknown as UnknownRecord,
+        ...newState,
+      });
+    }
+  }
+  const windowKey = placer(creator).windowKey;
+  dispatch(
```
Comment (emilbon99) [inline console/src/schematic/services/file.ts:72]: You should be able to set the mosaic tab location using the `placer`. There's no need for a separate call to `moveMosaicTab`

File: console/src/schematic/services/ontology.tsx [hunk 8]
```diff
@@ -117,6 +120,33 @@ const useRangeSnapshot = (): ((props: Ontology.TreeContextMenuProps) => void) =>
...
+const useDownload = (): ((props: Ontology.TreeContextMenuProps) => void) =>
+  useMutation<void, Error, Ontology.TreeContextMenuProps, Tree.Node[]>({
+    mutationFn: async ({ client, selection: { resources }, store }) => {
+      let state = select(store.getState(), resources[0].id.key);
+      if (state == null) {
+        const schematic = await client.workspaces.schematic.retrieve(
+          resources[0].id.key,
+        );
+        state = schematic.data as unknown as State;
+        state.snapshot = schematic.snapshot;
```
Comment (emilbon99) [inline console/src/schematic/services/ontology.tsx:132]: Should we preserve the snapshot field when exporting schematics? I don't think so.

File: pluto/src/vis/diagram/edge/connector/connector.ts [hunk 9]
```diff
@@ -118,6 +119,11 @@ export interface Segment {
   length: number;
 }
 
+export const segmentZ = z.object({
```
Comment (emilbon99) [inline pluto/src/vis/diagram/edge/connector/connector.ts:122]: If you're going to define a zod type for this, make the segment interface this instead:

`type Segment = z.infer<typeof segmentZ>`

File: pluto/src/vis/diagram/types.ts [hunk 10]
```diff
@@ -56,6 +67,15 @@ export interface Edge extends RFEdgeData {
...
+  key: z.string(),
+  source: z.string(),
+  target: z.string(),
+  selected: z.boolean(),
+  sourceHandle: z.string().nullable(),
+  targetHandle: z.string().nullable(),
+});
+
 /** The properties for a node within a diagram. */
 export interface Node {
```
Comment (emilbon99) [inline pluto/src/vis/diagram/types.ts:80]: make this a z.infer

File: pluto/src/vis/diagram/types.ts [hunk 11]
```diff
@@ -26,6 +27,11 @@ export interface Viewport {
   zoom: number;
 }
 
+export const viewportZ = z.object({
+  position: xy.xy,
+  zoom: z.number(),
+});
+
 /** Pluto specific info passed to the 'data' attribute on rf.Edge */
 export interface RFEdgeData {
```
Comment (emilbon99) [inline pluto/src/vis/diagram/types.ts:36]: Make this a z.infer

File: pluto/src/vis/diagram/types.ts [hunk 12]
```diff
@@ -34,6 +40,11 @@ export interface RFEdgeData {
   segments: connector.Segment[];
 }
 
+export const rfEdgeDataZ = z.object({
+  color: Color.crudeZ,
+  segments: z.array(connector.segmentZ),
+});
+
 /** The properties for an edge within a diagram. */
 export interface Edge extends RFEdgeData {
```
Comment (emilbon99) [inline pluto/src/vis/diagram/types.ts:49]: Make this a z.infer

**Review by emilbon99 — CHANGES_REQUESTED** [review]
Looks mostly good. Take a look at my comments.

One major thing we should add . Maybe its in this PR, maybe its in a different one, but you should be able to import a schematic via the command palette

**Review by Lham42 — APPROVED** [review]

**Pull request merged**